raiden/network/proxies/user_deposit.py

```diff
@@ -236,118 +256,204 @@ def _init(
+            # reasonable value.
+            previous_total_deposit = self.get_total_deposit(
+                address=beneficiary, block_identifier=self.client.get_checking_block()
+            )
+            amount_to_deposit = TokenAmount(total_deposit - previous_total_deposit)
+        except BadFunctionCallOutput:
+            raise_on_call_returned_empty(given_block_identifier)
+        else:
+            amount_to_deposit = TokenAmount(total_deposit - previous_total_deposit)
+
+            if whole_balance + amount_to_deposit > UINT256_MAX:
+                msg = (
+                    f"Current whole balance is {whole_balance}. "
+                    f"The new deposit of {amount_to_deposit} would lead to an overflow."
+                )
+                raise BrokenPreconditionError(msg)
+
+            if whole_balance + amount_to_deposit > whole_balance_limit:
+                msg = (
+                    f"Current whole balance is {whole_balance}. "
+                    f"With the new deposit of {amount_to_deposit}, the deposit "
+                    f"limit of {whole_balance_limit} would be exceeded."
+                )
+                raise BrokenPreconditionError(msg)
+
+            if total_deposit <= previous_total_deposit:
+                msg = (
+                    f"Current total deposit {previous_total_deposit} is already larger "
+                    f"than the requested total deposit amount {total_deposit}"
+                )
+                raise BrokenPreconditionError(msg)
+
+            if current_balance < amount_to_deposit:
+                msg = (
+                    f"new_total_deposit - previous_total_deposit = {amount_to_deposit} "
+                    f"can not be larger than the available balance {current_balance}, "
+                    f"for token at address {to_checksum_address(token.address)}"
+                )
+                raise BrokenPreconditionError(msg)
+
+        return previous_total_deposit, amount_to_deposit
+
+    @contextmanager
+    def _deposit_inflight(
+        self, beneficiary: Address, total_deposit: TokenAmount
+    ) -> Iterator[None]:
+        """ Updates the `_inflight_deposits` dictionary to handle concurrent deposits.
+
+        Note: This must be called after `_deposit_preconditions`.
+        """
+        async_result = AsyncResult()
+        current_inflight = InflightDeposit(total_deposit, async_result)
+        self._inflight_deposits[beneficiary] = current_inflight
+
+        try:
+            yield
+        except Exception as e:
+            async_result.set_exception(e)
+            raise
+        else:
+            async_result.set_result(None)
+        finally:
+            # Clearing the dictionary is not strinctly necessary since the
+            # deposit is always increasing. But it is just nice to clear up the
+            # container.
+            del self._inflight_deposits[beneficiary]
```

Comment:
What happens when we have two inflight deposits for the same beneficiary (the second has a higher value than the previous one)?

Reply:
hummm, good catch, this should only be cleared if the objects are equal